### PR TITLE
ref: Redefine Replacement as ABC

### PR DESCRIPTION
**Pull request by untitaker**
This change has been motivated by two comments:

1. https://github.com/getsentry/snuba/pull/1476#discussion_r521151540
2. https://github.com/getsentry/snuba/pull/1795#discussion_r606420996

<details><summary>Old description</summary>
...though the interface may have to change in a bunch of ways to
actually solve the problems that Filippo complains about wrt poor typing
of queries. The important bit is that replacements now encapsulate their
own logic.

The old definition of replacement is implemented in terms of the new
one, and acts as a compatibility layer until all replacements have been
migrated.

Only two replacements have been actually refactored. exclude_groups has
been refactored to be a "modern" replacement, replace_groups has been
refactored just to be a type.

If we are fine with the Replacement interface, we can gradually:

1. refactor all replacements to look like ReplaceGroupReplacement
   (direct subtypes of LegacyReplacement)
2. get rid of LegacyReplacement

</details>

### CODE REVIEW

File: snuba/replacers/replacer_processor.py [hunk 1]
```diff
@@ -16,14 +17,36 @@ class ReplacementMessage(NamedTuple):
...
+    def get_insert_query_template(self) -> Optional[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_count_query_template(self) -> Optional[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_query_args(self) -> Mapping[str, Any]:
+        raise NotImplementedError()
```
Comment (fpacifici) [inline snuba/replacers/replacer_processor.py:49]: Is this the part you were referring to as 

> ...though the interface may have to change in a bunch of ways to
actually solve the problems that Filippo complains about wrt poor typing
of queries. The important bit is that replacements now encapsulate their
own logic.

I think we could get rid of this template entirely and let the instances of this class produce a query given some parameters provided by the replacer processor. But I would be ok in doing that in a second step and starting by reorganizing the code.

File: snuba/replacers/replacer_processor.py [hunk 2]
```diff
@@ -16,14 +17,36 @@ class ReplacementMessage(NamedTuple):
...
-    # replacement without running any query.
-    count_query_template: Optional[str]
-    insert_query_template: Optional[str]
-    query_args: Mapping[str, Any]
-    query_time_flags: Any
+class Replacement(ABC, Generic[C]):
+    @abstractclassmethod
+    def parse_message(
+        cls, message: Mapping[str, Any], context: C
+    ) -> Optional["Replacement[C]"]:
```
Comment (fpacifici) [inline snuba/replacers/replacer_processor.py:24]: I am not sure I get why you need to make this class generic to the type of context.  It seems you do not care about that type in any of the interface method except for the parser. So it seems the type of context the class operates on could be an implementation detail of the concrete classes with no need for this to be generic. As a consequence you would not have to reference the specific parameter type every time you reference `Replacement`.

Also you do not need to specify the `parse_message` method in the abstract class. The code that instantiates the Replacement object knows the concrete classes. It can reference the specific parse_message method.

**Review by fpacifici — COMMENTED** [review]
Thanks I think this goes in a good direction. A few high level comments

File: snuba/replacers/replacer_processor.py (hunk 2 above)
Comment (untitaker) [reply snuba/replacers/replacer_processor.py:24]: You are right, if I remove `parse_message` from the ABC I have removed all exposure of context to `replacer_processor`.

I may have to write an additional ABC for errors_replacer, put parse_message there and also get_excluded_groups as both are errors_replacer specific

File: snuba/replacers/replacer_processor.py (hunk 1 above)
Comment (untitaker) [reply snuba/replacers/replacer_processor.py:49]: could you elaborate what you want to do here? specifically, how do you want to communicate data between ReplacerWorker that needs those attrs and ReplacerProcessor implementors

File: snuba/datasets/errors_replacer.py [hunk 3]
```diff
@@ -176,37 +245,42 @@ def process_message(self, message: ReplacementMessage) -> Optional[Replacement]:
...
         else:
             raise InvalidMessageType("Invalid message type: {}".format(type_))
 
         return processed
 
-    def pre_replacement(self, replacement: Replacement, matching_records: int) -> bool:
-        # query_time_flags == (type, project_id, [...data...])
-        flag_type, project_id = replacement.query_time_flags[:2]
+    def pre_replacement(  # type: ignore[override]
+        self, replacement: Replacement, matching_records: int
```
Comment (untitaker) [inline snuba/datasets/errors_replacer.py:257]: liskov substitution principle!

what I actually want here is an associated type, like in Rust:

```
impl ReplacementProcessor for ErrorsReplacer {
    type Replacement = ErrorReplacement;

    fn pre_replacement(&self, replacement: Self::Replacement) -> ...
    
}
```

alternatively I can probably make it work by putting a generic on ReplacementProcessor like I did with Replacement in prev commit

File: snuba/replacers/replacer_processor.py (hunk 1 above)
Comment (fpacifici) [reply snuba/replacers/replacer_processor.py:35]: This is what we do today:
https://github.com/getsentry/snuba/blob/master/snuba/replacer.py#L53-L62

We return a template + a dictionary of arguments to replace in the template. The arguments are `replacement.query_args`. Some arguments are populated by replacer.py like `table_name`.

This creates a lot of coupling between the two classes. replacer.py relies on the template structure and replacer_processor relies on replacer.py to fill in a missing parameter. All this is not typed.

What I think we could do is to make the `Replacement` class provide methods that just return the SQL query:

```
class Replacer(ABC):
   def get_count_query(self, table_name: str):
       ....

   def get_insert_query(self, table_name: str):
       ....
```

So only the Replacement class will deal with the template and will never expose it. This makes this logic easy to test and reduces coupling. 

Though it makes sense to me to do such change in a separate PR to reduce the scope

File: snuba/datasets/errors_replacer.py [hunk 4]
```diff
@@ -38,6 +43,66 @@ class ReplacerState(Enum):
...
+    @abstractmethod
+    def get_needs_final(self) -> bool:
+        raise NotImplementedError()
+
+
+EXCLUDE_GROUPS = object()
+NEEDS_FINAL = object()
+QueryTimeFlags = Union[Tuple[object, int], Tuple[object, int, Any]]
+
+# Separate mixin because of https://github.com/python/mypy/issues/5374
```
Comment (fpacifici) [inline snuba/datasets/errors_replacer.py:72]: Not sure I get this. 
Legacy Replacement is not abstract, you should be able to make it a dataclass. without having a separate mixin class.
We do it often for Expression https://github.com/getsentry/snuba/blob/master/snuba/query/expressions.py#L116-L117

File: snuba/datasets/errors_replacer.py (hunk 3 above)
Comment (fpacifici) [reply snuba/datasets/errors_replacer.py:257]: > alternatively I can probably make it work by putting a generic on ReplacementProcessor like I did with Replacement in prev commit

I think the generic idea would work and be clean. `ReplacementProcessor` only defines the interface based on the `Repalcement` class, so you can easily make it a parameter. You can make it bounded to the Replacement class but that would not be required.

**Comment by untitaker** [comment]
Only need to port delete_tags replacement and then it's ready for final review

**Comment by untitaker** [comment]
It is done. If we think this is too risky I believe we can review & deploy this almost commit-by-commit.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Thanks for doing this.
Though I think this is way too big (read risky) to be in one single PR, as you suspected in your last comment.

I think we should find a strategy to split this PR. 
Just a few suggestions: 
- I liked the previoius version where you used one or two replacement messages as proofs of concepts. That was easy and comfortable to review and not very risky. 
- As I mentioned before addressing the template coupling issue in a separate PR would be totally fine. 
- Another alternative would be to make the Replacement class to carry the `get_insert_query`,   `get_count_query` and the old template at the same time so we can still fully migrate one replacement type here and let the old ones work without changes.

Also one of the goals of breaking replacements into smaller classes is that we can more effectively test them. They should have unit tests as well now so we can cover corner cases we did not before.

Happy to chat to find a manageable strategy to review this step by step.

File: snuba/replacers/replacer_processor.py [hunk 5]
```diff
@@ -16,17 +15,20 @@ class ReplacementMessage(NamedTuple):
...
-class Replacement:
-    # XXX: For the group_exclude message we need to be able to run a
-    # replacement without running any query.
-    count_query_template: Optional[str]
-    insert_query_template: Optional[str]
-    query_args: Mapping[str, Any]
-    query_time_flags: Any
+class Replacement(ABC):
+    @abstractmethod
+    def get_insert_query(self, table_name: str) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (untitaker) [reply snuba/replacers/replacer_processor.py:20]: There is an exclude_groups "replacement" which only sets the exclusion key in Redis and does not need to touch ClickHouse. It's part of reprocessing.

File: snuba/datasets/storage.py [hunk 6]
```diff
@@ -167,7 +167,7 @@ def __init__(
         stream_loader: KafkaStreamLoader,
         query_splitters: Optional[Sequence[QuerySplitStrategy]] = None,
         mandatory_condition_checkers: Optional[Sequence[ConditionChecker]] = None,
-        replacer_processor: Optional[ReplacerProcessor] = None,
+        replacer_processor: Optional[ReplacerProcessor[Any]] = None,
```
[the remark this answers is not in the document]
Comment (untitaker) [reply snuba/datasets/storage.py:170]: `ErrorsReplacer` is not a valid substitute for `ReplacerProcessor[Replacement]` because from `issubclass(Foo, Bar)` it does not follow that `issubclass(ReplacerProcessor[Foo], ReplacerProcessor[Bar])`.

The core issue is that there are two `Replacement` classes, one that only contains data relevant to `ReplacerWorker`, and a subclass that contains additional attributes needed by `ErrorsReplacer`.

File: snuba/datasets/errors_replacer.py [hunk 7]
```diff
@@ -1,43 +1,64 @@
...
+
+
+class Replacement(ReplacementBase):
+    @abstractmethod
+    def get_excluded_groups(self) -> Sequence[int]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_needs_final(self) -> bool:
+        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (untitaker) [reply snuba/datasets/errors_replacer.py:55]: See https://github.com/getsentry/snuba/pull/1861#discussion_r632427814

We can merge the two Replacement classes but I believe that needs_final is specific to ErrorsReplacer.

File: snuba/datasets/errors_replacer.py [hunk 8]
```diff
@@ -144,323 +165,333 @@ def __init__(
...
         if not settings.REPLACER_IMMEDIATE_OPTIMIZE:
-            if flag_type == NEEDS_FINAL:
+            if needs_final:
                 if compatibility_double_write:
                     set_project_needs_final(project_id, None)
                 set_project_needs_final(project_id, self.__state_name)
-            elif flag_type == EXCLUDE_GROUPS:
-                group_ids = replacement.query_time_flags[2]
+
+            if group_ids_to_exclude:
```
[the remark this answers is not in the document]
Comment (untitaker) [reply snuba/datasets/errors_replacer.py:212]: Do you have a suggestion for how to make the ABC interface reflect that? I am not super comfortable with allowing `Replacement` to be implemented like that while disallowing this behavior at runtime.

[the remark this answers is not in the document]
Comment (untitaker) [reply snuba/datasets/errors_replacer.py:212]: This particular concern should be addressed in https://github.com/getsentry/snuba/pull/1877 in some way, not sure if that's a nice API but it sure is typesafe now

**Comment by fpacifici** [comment]
Let's resurrect it when we get back to this. 
I think we may have to rewrite quite a bit of it anyways as the code structure has changed.
